### PR TITLE
fix: clean up dangling buckets during bucket delete

### PR DESCRIPTION
**Pull request by poornas**
If all disks are online and report volume not found
or volume not empty, it should be safe to treat this
as stale bucket during DeleteBucket call.

## How to test this PR?
With dangling bucket in namespace like below, `mc rb` should succeed
 /tmp tree
.
├── export1
│   └── processing
│       └── issue
│           └── xl.meta
├── export2
├── export3
├── export4
➜  /tmp mc rb minio/processing --force
mc: <ERROR> Failed to remove `minio/processing`. The specified bucket does not exist.

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

File: cmd/erasure-bucket.go [hunk 1]
```diff
@@ -172,7 +172,7 @@ func (er erasureObjects) DeleteBucket(ctx context.Context, bucket string, opts D
 	if err == errErasureWriteQuorum && !opts.NoRecreate {
 		undoDeleteBucket(storageDisks, bucket)
 	}
-	if err != nil {
+	if err != nil && !isStaleBucket(dErrs) {
```
Comment (vadmeste) [inline cmd/erasure-bucket.go]: This feels dangerous, isStaleBucket() returns true if all buckets are not empty. Maybe you want to add `&& opts.Force` here.

**Review by vadmeste — COMMENTED** [review]
One comment

**Comment by klauspost** [comment]
Test failure:
```
2021-10-27T23:19:02.8387614Z === RUN   TestServerSuite/Test:_5,_ServerType:_ErasureSet
2021-10-27T23:19:02.8388271Z Formatting 1st pool, 2 set(s), 16 drives per set.
2021-10-27T23:19:02.8389038Z WARNING: Host local has more than 4 drives of set. A host failure will result in data becoming unavailable.
2021-10-27T23:19:02.8389991Z WARNING: Host local has more than 4 drives of set. A host failure will result in data becoming unavailable.
2021-10-27T23:19:02.8391135Z     server_test.go:61: Test [server_test.go:456:(*TestSuiteCommon).TestDeleteBucketNotEmpty()]:ErasureSet expected 409, got 204
2021-10-27T23:19:02.8392071Z --- FAIL: TestServerSuite (62.04s)
2021-10-27T23:19:02.8392702Z     --- PASS: TestServerSuite/Test:_1,_ServerType:_FS (9.13s)
2021-10-27T23:19:02.8393433Z     --- PASS: TestServerSuite/Test:_2,_ServerType:_FS (9.00s)
2021-10-27T23:19:02.8394150Z     --- PASS: TestServerSuite/Test:_3,_ServerType:_FS (41.68s)
2021-10-27T23:19:02.8394883Z     --- FAIL: TestServerSuite/Test:_4,_ServerType:_Erasure (1.01s)
2021-10-27T23:19:02.8395689Z     --- FAIL: TestServerSuite/Test:_5,_ServerType:_ErasureSet (1.22s)
```

**Comment by harshavardhana** [comment]
This kind of deletes are not correct @poornas you need to have pool level awareness here.

Just doing the checks at the lowest layer will not have context leading to more partial buckets.

**Comment by harshavardhana** [comment]
The correct fix as discussed @poornas 

```diff
diff --git a/cmd/erasure-bucket.go b/cmd/erasure-bucket.go
index 7222ac2e7..333a3687d 100644
--- a/cmd/erasure-bucket.go
+++ b/cmd/erasure-bucket.go
@@ -172,19 +172,23 @@ func (er erasureObjects) DeleteBucket(ctx context.Context, bucket string, opts D
        if err == errErasureWriteQuorum && !opts.NoRecreate {
                undoDeleteBucket(storageDisks, bucket)
        }
-       if err != nil {
-               return toObjectErr(err, bucket)
-       }
-
-       // At this point we have `err == nil` but some errors might be `errVolumeNotEmpty`
-       // we should proceed to attempt a force delete of such buckets.
-       for index, err := range dErrs {
-               if err == errVolumeNotEmpty && storageDisks[index] != nil {
-                       storageDisks[index].RenameFile(ctx, bucket, "", minioMetaTmpDeletedBucket, mustGetUUID())
+       if err == nil || errors.Is(err, errVolumeNotFound) {
+               // At this point we have `err == nil` but some errors might be `errVolumeNotEmpty`
+               // we should proceed to attempt a force delete of such buckets.
+               var purgedDangling bool
+               for index, err := range dErrs {
+                       if err == errVolumeNotEmpty && storageDisks[index] != nil {
+                               storageDisks[index].RenameFile(ctx, bucket, "", minioMetaTmpDeletedBucket, mustGetUUID())
+                               purgedDangling = true
+                       }
+               }
+               // if we purged dangling buckets, ignore errVolumeNotFound errors.
+               if purgedDangling {
+                       err = nil
                }
        }
 
-       return nil
+       return toObjectErr(err, bucket)
 }
 
```

File: cmd/erasure-bucket.go [hunk 2]
```diff
@@ -172,19 +172,23 @@ func (er erasureObjects) DeleteBucket(ctx context.Context, bucket string, opts D
...
+		var purgedDangling bool
+		for index, err := range dErrs {
+			if err == errVolumeNotEmpty && storageDisks[index] != nil {
+				storageDisks[index].RenameFile(ctx, bucket, "", minioMetaTmpDeletedBucket, mustGetUUID())
+				purgedDangling = true
+			}
+			// if we purged dangling buckets, ignore errVolumeNotFound error.
+			if purgedDangling {
+				err = nil
+			}
```
Comment (harshavardhana) [inline cmd/erasure-bucket.go]: This can come out of the loop @poornas

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**